src/TestProblem/ELBDM/DiskHeating/Init_TestProb_ELBDM_DiskHeating.cpp

```diff
@@ -0,0 +1,1208 @@
+
+      // velocity dispersion table for thin disk
+      if ( MPI_Rank == 0 )
+      {
+         counter = 0;
+         input_table = fopen(DispTableFile,"r");
+         if (!input_table)
+         Aux_Error( ERROR_INFO, "File %s cannot be opened! Exit!\n", DispTableFile);
+         else
+            Aux_Message( stdout, "Loading %s for reading density table succeed!\n", DispTableFile);
+         while (!feof(input_table))
+         {
+            fgets(buff, STR_SIZE, input_table);
+            if (buff==NULL||buff[0]=='\0')
+               break;
+            else if (buff[0]!='#')
+            {
+               sscanf(buff, "%lf %lf", &radius, &value);
+               DispTable_r[counter] = radius;
+               DispTable_d[counter] = value*1e5/UNIT_V;
+               counter ++;
+            }
+            memset(buff, '\0', STR_SIZE);
+         }
+         fclose(input_table);
+      }
+      MPI_Bcast(DispTable_r, DispNbin, MPI_DOUBLE, 0, MPI_COMM_WORLD);
+      MPI_Bcast(DispTable_d, DispNbin, MPI_DOUBLE, 0, MPI_COMM_WORLD);
+
+      if ( MPI_Rank == 0 )
+      {
+         Aux_Message( stdout, "%s ...\n", __FUNCTION__ );
+
+         const double ParM = Disk_Mass / NPar_AllRank;
+         double Ran, RanR, RanV, R, Rold, f, f_, phi, RanVec[3];
+         Aux_Message(stdout, " Particle Mass = %13.7e\n", ParM) ;
+
+//       initialize the RNG
+         RNG = new RandomNumber_t( 1 );
+         RNG->SetSeed( 0, 1002 );
+
+         for ( long p = 0; p < NPar_AllRank; p++)
+         {
+            Time_AllRank[p] = Time[0];
+//          mass
+            Mass_AllRank[p] = ParM;
+//          label
+            Type_AllRank[p] = 3;
+            Label_AllRank[p] = 3;
+
+//          position
+            Ran  = RNG->GetValue( 0, 0.0, 1.0);
+            R = 1.0;
+            do
+            {
+               f = (1 + R) * exp(-R) + Ran - 1;
+               f_ = -R * exp(-R);
+               Rold = R;
+               R = R - f / f_;
+            }
+            while(fabs(R - Rold) / R > 1e-7);
+            RanR = Disk_R*R;
+            phi = 2*pi*RNG->GetValue( 0, 0.0, 1.0 );
+            RanVec[0] = RanR*cos(phi);
+            RanVec[1] = RanR*sin(phi);
+            RanVec[2] = 0;
+            for (int d = 0; d < 3; d++) Pos_AllRank[d][p] = RanVec[d] + Cen[d];
+
+//          velocity
+            double V_acc, sigma, r;
+            double RanV[3];
+            V_acc = sqrt(Ran*Disk_Mass/RanR);
+            Vel_AllRank[0][p] = - V_acc*sin(phi);
+            Vel_AllRank[1][p] =   V_acc*cos(phi);
+            Vel_AllRank[2][p] =   0;
+
+         } // for ( long p = 0; p < NPar_AllRank; p++)
+
+      } //if ( MPI_Rank == 0 )
+   } //if ( !AddParWhenRestartByfile )
+
+// add particles here
+   Par_AddParticleAfterInit( NNewPar, NewParAtt );
+// free memory
+   for (int v=0; v<PAR_NATT_TOTAL; v++)   delete [] NewParAtt[v];
+
+// initialize the k-space Green's function for the isolated BC.
+   if ( OPT__BC_POT == BC_POT_ISOLATED )  Init_GreenFuncK();
```

Comment:
- This is part of the Poisson solver. So why not put it within `if ( !AddParWhenRestartByfile )` before calling `Poi_GetAverageDensity`, similar to the procedure in `Init_GAMER.cpp`?
- I'm worried it may cause memory leakage since `Init_GreenFuncK()`, which will allocate memory for `GreenFuncK`, will also be called in `Init_GAMER.cpp`. A temporary solution will be freeing this array by `root_fftw::fft_free(GreenFuncK); GreenFuncK      = NULL;` (as done in `End_MemFree_PoissonGravity.cpp`) after `Par_AfterAcceleration_Ptr()`. What do you think?